crates/runtime/src/externref.rs

```diff
@@ -757,176 +756,55 @@ struct StackMapRegistryInner {
     ranges: BTreeMap<usize, ModuleStackMaps>,
 }
 
-#[derive(Debug)]
 struct ModuleStackMaps {
-    /// The range of PCs that this module covers. Different modules must always
-    /// have distinct ranges.
-    range: std::ops::Range<usize>,
-
-    /// A map from a PC in this module (that is a GC safepoint) to its
-    /// associated stack map. If `None` then it means that the PC is the start
-    /// of a range which has no stack map.
-    pc_to_stack_map: Vec<(usize, Option<Rc<StackMap>>)>,
+    start: usize,
+    lookup: Box<dyn StackMapLookup>,
+}
+
+/// A trait used to abstract how stack maps are located.
+pub trait StackMapLookup {
```

Comment:
Question: why the trait over using a method on `CompiledModule` (similar to `func_by_pc` and `func_info`)?

Reply:
Only because the `CompiledModule` lives in the `wasmtime_jit` crate which this `wasmtime_runtime` crate doesn't have access to. Otherwise though there's definitely no need for dynamic trait dispatch here.

Reply:
Makes sense.  I guess the alternative would be to use a "lookup" trait that the runtime interacts with (it would just need the `lookup_stack_map` method), freeing the registry implementation to live in the `wasmtime` crate, but that's just trading dynamic-dispatch during module instantiation to dynamic-dispatch at externref gc.

Reply:
If we did move the registry implementation to `wasmtime`, I think this data structure could be a generic one replacing `StoreFrameInfo` too, basically just a "which `CompiledModule` should I be talking to given this pc?" which responds with an answer that we then use to lookup the appropriate information (gimmie frame info, gimmie trap info, gimmie stack maps, etc.)

Reply:
In fact, given externref gc being a relatively expensive (and hopefully very infrequent) operation and the cost of doing trait object dispatch per frame would be negligible, I would argue `VMContext` should have `*const dyn StackMapLookup` rather than `*mut StackMapRegistry`.

This frees us to remove `StackMapRegistry` and `StoreFrameInfo` in favor of a single `CompiledModuleRegistry` per store which impls the runtime's `StackMapLookup` and can lookup frame and trap information  as well.

Thoughts on this approach? Perhaps for another PR?

Reply:
Hm good points! I'll look to add this to this PR, seems reasonable to do here!